### PR TITLE
feat: [P4PU-562] adding whitelist

### PR DESCRIPTION
**Pull request by oleksiybozhykntt**
#### List of Changes
Added whitelist feature 

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 1]
```diff
@@ -20,28 +21,59 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
   private final JWTConfiguration jwtConfiguration;
+  private final AuthService authService;
 
   CustomAuthenticationSuccessHandler(
       AccessTokenBuilderService accessTokenBuilderService,
       ObjectMapper objectMapper,
       JWTConfiguration jwtConfiguration,
-      TokenStoreService tokenStoreService){
+      TokenStoreService tokenStoreService,
+      AuthService authService){
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:31]: the cf list has not usage outside of this component...
I will not expose it adding a new method on the AuthService...
let's just inject here the Set

ps: let's explicitly add the `white-list` property, bind to the ENV var, inside the application.yml
```suggestion
      @Value("${white-list}") Set<String> cfWhitelist){
```

**Review by antonioT90 — CHANGES_REQUESTED** [review]
let's make simpler the solution by modifying just the AuthenticationHandler... the Service change is not required, because the list has not other use than the handler

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 2]
```diff
@@ -35,13 +40,41 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
+      authentication.setAuthenticated(false);
+      response.setStatus(403);
+      response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+      response.getWriter().write("{\"error\": \"User not allowed to access this application\"}");
+      response.getWriter().flush();
+    }
+
+  }
+
+  boolean isInWhiteList(String fiscalCode){
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:70]:
```suggestion
  private boolean isInWhiteList(String fiscalCode){
```

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 3]
```diff
@@ -35,13 +40,41 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
+    if (whiteListUsers == null) {
+      return false; // or handle the null case as needed
+    }
 
-    TokenResponse accessToken = new TokenResponse(accessTokenBuilderService.build(),jwtConfiguration.getTokenType(),jwtConfiguration.getAccessToken().getExpireIn(),null,null);
-    tokenStoreService.save(accessToken.getAccessToken(), IamUserInfoDTO.map2IamUserInfoDTO(oauth2Token.getPrincipal().getAttributes()));
-    response.setContentType(MediaType.APPLICATION_JSON_VALUE);
-    response.getWriter().write( objectMapper.writeValueAsString(accessToken) );
-    response.getWriter().flush();
+    return whiteListUsers.stream().anyMatch(str -> str.equals(fiscalCode));
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:77]: use Set contains method, it's not necessary to use the streams API

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 4]
```diff
@@ -35,13 +40,41 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
+    }
+
+  }
+
+  boolean isInWhiteList(String fiscalCode){
+    Set<String> whiteListUsers = cfWhitelist;
+
+    if (whiteListUsers == null) {
+      return false; // or handle the null case as needed
+    }
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:75]: this is not something to check! the property is mandatory

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 5]
```diff
@@ -35,13 +40,41 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
+      response.setStatus(403);
+      response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+      response.getWriter().write("{\"error\": \"User not allowed to access this application\"}");
+      response.getWriter().flush();
+    }
+
+  }
+
+  boolean isInWhiteList(String fiscalCode){
+    Set<String> whiteListUsers = cfWhitelist;
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:71]: this is useless

File: src/main/resources/application.yml [hunk 6]
```diff
@@ -21,6 +21,7 @@ spring:
             token-uri: \${AUTH_CLIENT_TOKEN_URI:}
             user-name-attribute: sub
             jwk-set-uri: \${AUTH_CLIENT_JWK_URI:}
+white-list: \${WHITE_LIST_USERS:}
```
Comment (antonioT90) [inline src/main/resources/application.yml:24]: can you give to this property a more significant name? this is a white list of what?

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/resources/application.yml (hunk 6 above)
Comment (oleksiybozhykntt) [reply src/main/resources/application.yml]: renamed

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java (hunk 5 above)
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java]: removed

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java (hunk 4 above)
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java]: ok

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java (hunk 2 above)
Comment (oleksiybozhykntt) [reply src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java]: ok

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 7]
```diff
@@ -21,6 +23,9 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
   private final TokenStoreService tokenStoreService;
   private final JWTConfiguration jwtConfiguration;
 
+  @Value("${white-list-users}")
+  private Set<String> cfWhitelist;
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:27]: use always constructor injection!

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java [hunk 8]
```diff
@@ -35,13 +40,35 @@ public class CustomAuthenticationSuccessHandler implements AuthenticationSuccess
...
+      response.getWriter().write( objectMapper.writeValueAsString(accessToken) );
+      response.getWriter().flush();
+    } else {
+      authentication.setAuthenticated(false);
+      response.setStatus(403);
+      response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+      response.getWriter().write("{\"error\": \"User not allowed to access this application\"}");
+      response.getWriter().flush();
+    }
+
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:67]: as you can see, you are repeating some code! let's make it more simples
```suggestion
      body=  objectMapper.writeValueAsString(accessToken);
    } else {
      authentication.setAuthenticated(false);
      response.setStatus(403);
      body="{\"error\": \"User not allowed to access this application\"}";
    }
    
      response.setContentType(MediaType.APPLICATION_JSON_VALUE);
      response.getWriter().write( body );
      response.getWriter().flush();

```

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -21,6 +21,7 @@ spring:
             token-uri: \${AUTH_CLIENT_TOKEN_URI:}
             user-name-attribute: sub
             jwk-set-uri: \${AUTH_CLIENT_JWK_URI:}
+white-list-users: \${WHITE_LIST_USERS:}
```
Comment (antonioT90) [inline src/main/resources/application.yml:24]: use semantic names that help to recognize also the content!
```suggestion
white-list-cf-users: \${WHITE_LIST_CF_USERS:}
```

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java [hunk 10]
```diff
@@ -34,15 +38,50 @@ public class CustomAuthenticationSuccessHandlerTest {
...
+
+    customAuthenticationSuccessHandler.onAuthenticationSuccess(request,response,oAuth2AuthenticationToken);
+
+    assertEquals(ContentType.APPLICATION_JSON.getMimeType(),response.getContentType());
+
+    TokenResponse token = new ObjectMapper().readValue(response.getContentAsString(),TokenResponse.class);
+    assertNotNull(token.getAccessToken());
+    assertNotNull(tokenStoreService.getUserInfo(token.getAccessToken()));
+    assertNotNull(token.getTokenType());
+    assertNotNull(token.getExpiresIn());
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java:63]: can you assert values instead that just not null?

File: src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java [hunk 11]
```diff
@@ -34,15 +38,50 @@ public class CustomAuthenticationSuccessHandlerTest {
...
+      throws IOException {
+    OAuth2AuthenticationToken oAuth2AuthenticationToken = getAuthenticationToken();
+
+    MockHttpServletResponse response = new MockHttpServletResponse();
+    MockHttpServletRequest request = new MockHttpServletRequest();
+
+    customAuthenticationSuccessHandler.onAuthenticationSuccess(request,response,oAuth2AuthenticationToken);
+
+    assertEquals(ContentType.APPLICATION_JSON.getMimeType(), response.getContentType());
+    assertEquals(403,response.getStatus());
```
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java:77]: assert also the body

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandler.java:1]: I think that this class is not a service (I mean the package, it's a security component

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java
Comment (antonioT90) [inline src/test/java/it/gov/pagopa/arc/service/CustomAuthenticationSuccessHandlerTest.java:1]: move also this class to the right package

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Review by antonioT90 — APPROVED** [review]

**Pull request merged**